includes/sanitizers/class-amp-gtag-script-sanitizer.php

```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * GA4 Script Sanitizer.
+ *
+ * - This sanitizer will facilitate using GA4 while waiting on an AMP implementation.
+ * - This sanitizer will be only used in Moderate or Loose sandboxing level
+ *
+ * @since 2.3.1
+ * @package AMP
+ */
+
+use AmpProject\AmpWP\Option;
+use AmpProject\AmpWP\ValidationExemption;
+
+/**
+ * Class AMP_GTag_Script_Sanitizer
+ *
+ * @since 2.3.1
+ * @internal
+ */
+class AMP_GTag_Script_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Sanitize the AMP response for GA4 scripts.
+	 *
+	 * @since 2.3.1
+	 */
+	public function sanitize() {
+		if ( ! AMP_Options_Manager::get_option( Option::SANDBOXING_ENABLED ) ) {
+			return;
+		}
+
+		$sandboxing_level = AMP_Options_Manager::get_option( Option::SANDBOXING_LEVEL );
+
+		if ( 1 !== $sandboxing_level && 2 !== $sandboxing_level ) {
+			return;
+		}
+
+		/**
+		 * GTag Script looks like this:
+		 *
+		 * <script async src="https://www.googletagmanager.com/gtag/js?id=xxxxxx"></script>
+		 * <script>
+		 *   window.dataLayer = window.dataLayer || [];
+		 *   function gtag(){dataLayer.push(arguments);}
+		 *   gtag('js', new Date());
+		 *
+		 *   gtag('config', 'xxxxxx');
+		 * </script>
+		 */
+		$scripts = $this->dom->xpath->query( '//script[ ( @async and starts-with( @src, "https://www.googletagmanager.com/gtag/js" ) ) or contains( text(), "function gtag(" ) ]' );
```

Comment:
```suggestion
		$scripts = $this->dom->xpath->query( '//script[ ( @async and starts-with( @src, "https://www.googletagmanager.com/gtag/js" ) ) or contains( text(), "gtag(" ) ]' );
```

@westonruter After updating the XPath query in https://github.com/ampproject/amp-wp/pull/7290/commits/b88130ff2770183fc83f8c13e35543e3e8d32cd8 we added `function gtag(`. It will not PX verify the custom scripts with `gtag` events which can look like:
```html
<script>
   // Send event
   gtag("event", "click", {
       "event_category": "social",
       "event_label": "twitter"
   });
   
   // Pageview
   gtag('event', 'page_view', {
       "page_title": "landing-page",
   });
</script>
```

Reply:
I haven't done much with gtag myself. Is this common? Are there `gtag()` calls in scripts other than the one in which the `gtag` function is defined?

Reply:
> Are there gtag() calls in scripts other than the one in which the gtag function is defined?

Not initially but devs like to create custom events. 

> Is this common?

I have seen it in many places and there are questions related to it on StackOverflow with high activitiy.

Reply:
Found such a use case on [this](https://getbootstrap.com/) site. They have multiple `ga` events on their page with onclick,

Reply:
I'm not finding `gtag` anywhere on the page source. Could you point it to me?

My only concern about marking all `script` tags containing `gtag` is it's possible there could be a lot of other JS in there which is not actually PX-verified. Maybe devs should manually add `data-px-verified-tag` to any _additional_ `script` tags beyond the one that Google provides?

Reply:
> I'm not finding gtag anywhere on the page source. Could you point it to me?

You can find `ga` which is now replaced by gtag in new version.

> My only concern about marking all script tags containing gtag is it's possible there could be a lot of other JS in there which is not actually PX-verified.

Yes it's possible. In that case we can add a note about adding PX verified attribute on gtag events scripts.

Reply:
Oh interesting. I see all of those `ga()` calls now. It turns out that none of them actually would be allowlisted if the XPath query were made to just look for `gtag` since all of the instances are in `onclick` attributes, not inline script tags. For them we'd have to do something else, to query all `//*/@onclick[ starts-with( ., 'gtag(' ) ]` to mark the _attributes_ as being PX-verified, which would result in `data-px-verified-attr="onclick"` being added as another attribute. There would likely be other such inline event handlers to account for as well, including `onsubmit` and `onkeypress`, etc. Such event handlers should only be queried if an initial `gtag` script was found. If you want to include that as part of this PR, then I'm happy with that.

Reply:
See also this existing query for all inline event handler attributes:

https://github.com/ampproject/amp-wp/blob/ccf43a32857f653a824efb0afbc219b9c3bdc73c/includes/sanitizers/class-amp-script-sanitizer.php#L300-L308

Reply:
Added in https://github.com/ampproject/amp-wp/pull/7290/commits/df3ee7b1fb9ffaf090df80a0f11cedd46dda7003.